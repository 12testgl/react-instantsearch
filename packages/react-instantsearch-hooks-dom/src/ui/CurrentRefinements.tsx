import React from 'react';

import { cx } from './lib/cx';

type CurrentRefinementCategory = Pick<
  React.ButtonHTMLAttributes<HTMLButtonElement>,
  'onClick'
> & { label: string };

type CurrentRefinement = {
  label: string;
  category: CurrentRefinementCategory[];
};

export type CurrentRefinementsProps = React.HTMLAttributes<HTMLDivElement> & {
  items?: CurrentRefinement[];
  hasRefinements?: boolean;
};

export function CurrentRefinements({
  items = [],
  hasRefinements = false,
  ...props
}: CurrentRefinementsProps) {
  return (
    <div
      {...props}
      className={cx(
        'ais-CurrentRefinements',
        !hasRefinements && 'ais-CurrentRefinements--noRefinement',
        props.className
      )}
    >
      <ul
        className={cx(
          'ais-CurrentRefinements-list',
          !hasRefinements && 'ais-CurrentRefinements-list--noRefinement'
        )}
      >
        {items.map((item) => (
          <li key={item.label} className="ais-CurrentRefinements-item">
            <span className="ais-CurrentRefinements-label">{item.label}:</span>
            {item.category.map(({ label, onClick }) => (
              <span key={label} className="ais-CurrentRefinements-category">
                <span className="ais-CurrentRefinements-categoryLabel">
                  {label}
                </span>
                <button
                  onClick={onClick}
                  className="ais-CurrentRefinements-delete"
                >
                  âœ•
                </button>
              </span>
            ))}
          </li>
        ))}
      </ul>
    </div>
  );
}
