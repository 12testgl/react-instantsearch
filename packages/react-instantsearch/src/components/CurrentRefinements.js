import PropTypes from 'prop-types';
import React, { Component } from 'react';

import translatable from '../core/translatable';

class CurrentRefinements extends Component {
  static propTypes = {
    cx: PropTypes.func.isRequired,
    items: PropTypes.arrayOf(
      PropTypes.shape({
        label: PropTypes.string,
      })
    ).isRequired,
    refine: PropTypes.func.isRequired,
    canRefine: PropTypes.bool.isRequired,
    transformItems: PropTypes.func,
    translate: PropTypes.func.isRequired,
  };

  render() {
    const { cx, translate, items, refine } = this.props;

    return (
      <ul className={cx('list')}>
        {items.filter(item => item.items).map(item =>
          item.items.map(nestedItem => (
            <li key={nestedItem.label} className={cx('item')}>
              <button
                className={cx('button')}
                onClick={() => refine(nestedItem.value)}
              >
                <span className={cx('label')}>{nestedItem.label}</span>
                <span className={cx('delete')}>
                  {translate('clearFilter', nestedItem)}
                </span>
              </button>
            </li>
          ))
        )}
      </ul>
    );
  }
}

export default translatable({
  clearFilter: 'âœ•',
})(CurrentRefinements);
