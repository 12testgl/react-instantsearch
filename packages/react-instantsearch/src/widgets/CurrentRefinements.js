import PropTypes from 'prop-types';
import React, { Component } from 'react';
import BaseWidget from './BaseWidget';
import connectCurrentRefinements from '../connectors/connectCurrentRefinements.js';
import CurrentRefinementsComponent from '../components/CurrentRefinements.js';
import classNames from '../components/classNames';

const cx = classNames('CurrentRefinements');

/**
 * The CurrentRefinements widget displays the list of currently applied filters.
 *
 * It allows the user to selectively remove them.
 * @name CurrentRefinements
 * @kind widget
 * @propType {function} [transformItems] - Function to modify the items being displayed, e.g. for filtering or sorting them. Takes an items as parameter and expects it back in return.
 * @propType {node} [header] - Adds a header to the widget.
 * @propType {node} [footer] - Adds a footer to the widget.
 * @themeKey ais-CurrentRefinements - the root div of the widget
 * @themeKey ais-CurrentRefinements-header - the header of the widget (optional)
 * @themeKey ais-CurrentRefinements-body - the body of the widget
 * @themeKey ais-CurrentRefinements-list - the list of all refined items
 * @themeKey ais-CurrentRefinements-item - the refined list item
 * @themeKey ais-CurrentRefinements-button - the button of each refined list item
 * @themeKey ais-CurrentRefinements-label - the refined list label
 * @themeKey ais-CurrentRefinements-count - the count of refined values for each item
 * @themeKey ais-CurrentRefinements-delete - the delete button of each label
 * @themeKey ais-CurrentRefinements-reset - the reset button for current selected values
 * @themeKey ais-CurrentRefinements-footer - the footer of the widget (optional)
 * @translationKey clearFilter - the remove filter button label
 * @example
 * import React from 'react';
 *
 * import { CurrentRefinements, RefinementList, InstantSearch } from 'react-instantsearch/dom';
 *
 * export default function App() {
 *   return (
 *     <InstantSearch
 *       appId="latency"
 *       apiKey="6be0576ff61c053d5f9a3225e2a90f76"
 *       indexName="ikea"
 *     >
 *       <CurrentRefinements />
 *       <RefinementList
          attributeName="colors"
          defaultRefinement={['Black']}
        />
 *     </InstantSearch>
 *   );
 * }
 */

class Widget extends Component {
  componentWillMount() {
    if (this.context.canRefine) this.context.canRefine(this.props.canRefine);
  }

  componentWillReceiveProps(props) {
    if (this.context.canRefine) this.context.canRefine(props.canRefine);
  }

  render() {
    const { header, footer, canRefine } = this.props;
    return (
      <BaseWidget
        cx={cx}
        header={header}
        footer={footer}
        cantRefine={!canRefine}
      >
        <CurrentRefinementsComponent cx={cx} {...this.props} />
      </BaseWidget>
    );
  }
}

Widget.propTypes = {
  canRefine: PropTypes.bool.isRequired,
  header: PropTypes.node,
  footer: PropTypes.node,
};

Widget.contextTypes = {
  canRefine: PropTypes.func,
};

export default connectCurrentRefinements(Widget);
