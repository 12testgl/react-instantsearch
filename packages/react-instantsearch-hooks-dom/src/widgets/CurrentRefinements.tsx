import React from 'react';
import { useCurrentRefinements } from 'react-instantsearch-hooks';

import { CurrentRefinements as CurrentRefinementsUiComponent } from '../ui/CurrentRefinements';

import { isModifierClick } from './lib/isModifierClick';

import type { CurrentRefinementsProps as CurrentRefinementsUiComponentProps } from '../ui/CurrentRefinements';
import type { UseCurrentRefinementsProps } from 'react-instantsearch-hooks';

export type CurrentRefinementsProps = Omit<
  CurrentRefinementsUiComponentProps,
  'disabled' | 'onClick' | 'items' | 'hasRefinements'
> &
  UseCurrentRefinementsProps;

export function CurrentRefinements(props: CurrentRefinementsProps) {
  const {
    includedAttributes,
    excludedAttributes,
    transformItems,
    ...otherProps
  } = props;
  const { items, refine, canRefine } = useCurrentRefinements(
    {
      includedAttributes,
      excludedAttributes,
      transformItems,
    },
    {
      $$widgetType: 'ais.currentRefinements',
    }
  );

  return (
    <CurrentRefinementsUiComponent
      {...otherProps}
      hasRefinements={canRefine}
      items={items.map((item) => ({
        label: item.label,
        categories: item.refinements.map((refinement) => ({
          label: refinement.label,
          onClick: (event) => {
            if (isModifierClick(event)) {
              return;
            }
            event.preventDefault();

            refine(refinement);
          },
        })),
      }))}
    />
  );
}
