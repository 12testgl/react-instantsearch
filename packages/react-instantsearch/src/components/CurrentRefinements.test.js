import PropTypes from 'prop-types';
/* eslint-env jest, jasmine */

import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({ adapter: new Adapter() });

import CurrentRefinements from './CurrentRefinements';

describe('CurrentRefinements', () => {
  it('renders a list of current refinements', () =>
    expect(
      renderer
        .create(
          <CurrentRefinements
            cx={(...x) => x.join(' ')}
            refine={() => null}
            items={[
              {
                label: 'Genre',
                value: 'clear all genres',
              },
            ]}
            canRefine={true}
          />
        )
        .toJSON()
    ).toMatchSnapshot());

  it('allows clearing unique items of a refinement', () =>
    expect(
      renderer
        .create(
          <CurrentRefinements
            cx={(...x) => x.join(' ')}
            refine={() => null}
            items={[
              {
                label: 'Genre',
                value: 'clear all genres',
                items: [
                  {
                    label: 'Sci-fi',
                    value: 'clear sci-fi',
                  },
                ],
              },
            ]}
            canRefine={true}
          />
        )
        .toJSON()
    ).toMatchSnapshot());
});
