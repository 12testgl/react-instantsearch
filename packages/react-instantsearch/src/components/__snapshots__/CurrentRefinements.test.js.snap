// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrentRefinements allows clearing unique items of a refinement 1`] = `
<div
  className="ais-CurrentRefinements"
>
  <ul
    className="ais-CurrentRefinements-list"
  >
    <li
      className="ais-CurrentRefinements-item"
    >
      <button
        className="ais-CurrentRefinements-button"
        onClick={[Function]}
      >
        <span
          className="ais-CurrentRefinements-label"
        >
          Sci-fi
        </span>
        <span
          className="ais-CurrentRefinements-delete"
        >
          ✕
        </span>
      </button>
    </li>
  </ul>
</div>
`;

exports[`CurrentRefinements renders a empty list with no current refinements 1`] = `
<div
  className="ais-CurrentRefinements ais-CurrentRefinements--noRefinement"
>
  <ul
    className="ais-CurrentRefinements-list ais-CurrentRefinements-list--noRefinement"
  />
</div>
`;

exports[`CurrentRefinements renders a list of current refinements 1`] = `
<div
  className="ais-CurrentRefinements"
>
  <ul
    className="ais-CurrentRefinements-list"
  >
    <li
      className="ais-CurrentRefinements-item"
    >
      <button
        className="ais-CurrentRefinements-button"
        onClick={[Function]}
      >
        <span
          className="ais-CurrentRefinements-label"
        >
          Genre
        </span>
        <span
          className="ais-CurrentRefinements-delete"
        >
          ✕
        </span>
      </button>
    </li>
  </ul>
</div>
`;
